File: app/src/main/java/com/darklycoder/wifitool/InputWiFiPasswordDialog.java
package com.darklycoder.wifitool;

import android.app.Dialog;
import android.content.Context;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import com.darklycoder.wifitool.lib.WiFiModule;
import com.darklycoder.wifitool.lib.info.WiFiScanInfo;

/**
 * wifi密码输入框
 */
public class InputWiFiPasswordDialog extends DialogFragment {

    TextView mTvTitle;
    EditText mEtPassword;
    TextView mTvConnect;
    ImageView mIvClose;
    TextView mTvConnectState;

    private WiFiScanInfo mInfo;
    private boolean isConnecting = false;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        getDialog().requestWindowFeature(Window.FEATURE_NO_TITLE);
        return inflater.inflate(R.layout.dialog_input_wifi_password, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        Bundle bundle = getArguments();
        if (null != bundle) {
            mInfo = (WiFiScanInfo) bundle.getSerializable("info");
        }
        if (null == mInfo) {
            dismissAllowingStateLoss();
            return;
        }

        mTvTitle = view.findViewById(R.id.tv_title);
        mEtPassword = view.findViewById(R.id.et_password);
        mTvConnect = view.findViewById(R.id.tv_connect);
        mIvClose = view.findViewById(R.id.iv_close);
        mTvConnectState = view.findViewById(R.id.tv_connect_state);

        mTvTitle.setText(mInfo.scanResult.SSID);

        mEtPassword.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                refreshConnectBtn();
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        mEtPassword.setFocusable(true);
        mEtPassword.setFocusableInTouchMode(true);
        mEtPassword.requestFocus();

        mEtPassword.post(new Runnable() {
            @Override
            public void run() {
                try {
                    InputMethodManager inputMethodManager = (InputMethodManager) getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
                    inputMethodManager.toggleSoftInput(0, InputMethodManager.HIDE_NOT_ALWAYS);

                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });

        mIvClose.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mEtPassword.setText(null);
            }
        });

        view.findViewById(R.id.tv_cancel).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                dismissAllowingStateLoss();
            }
        });

        view.findViewById(R.id.tv_connect).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //连接
                if (isConnecting) {
                    return;
                }

                isConnecting = true;
                WiFiModule.getInstance().connectWiFi(mInfo.scanResult.SSID, mInfo.getCipherType(), mEtPassword.getText().toString().trim());
                mTvConnect.setEnabled(!isConnecting);
            }
        });
    }

    @Override
    public void onStart() {
        super.onStart();

        Dialog dialog = getDialog();
        dialog.setCanceledOnTouchOutside(true);

        Window window = dialog.getWindow();
        if (null != window) {
            WindowManager.LayoutParams lp = window.getAttributes();
            lp.gravity = Gravity.CENTER;
            lp.width = dp2px(getContext(), 280);
            lp.height = WindowManager.LayoutParams.WRAP_CONTENT;
            window.setAttributes(lp);
        }
    }

    private boolean isEmpty(String password) {
        return TextUtils.isEmpty(password);
    }

    private boolean isValidPassword(String password) {
        if (isEmpty(password)) {
            return false;
        }

        //6~8位
        int size = password.length();
        if (size < 8) {
            return false;
        }

        return true;
    }

    private void refreshConnectBtn() {
        String password = mEtPassword.getText().toString().trim();

        mIvClose.setVisibility(isEmpty(password) ? View.GONE : View.VISIBLE);
        mTvConnect.setEnabled(!isConnecting && isValidPassword(password));
    }

    /**
     * 连接状态回调
     */
    public void connectCallBack(NetworkInfo.DetailedState state) {
        switch (state) {
            case CONNECTING:
                mTvConnectState.setText("连接中...");
                mTvConnectState.setVisibility(View.VISIBLE);
                break;

            case DISCONNECTED:
                isConnecting = false;
                refreshConnectBtn();
                mTvConnectState.setText("连接失败");
                mTvConnectState.setVisibility(View.VISIBLE);
                break;

            case CONNECTED:
                isConnecting = false;
                mTvConnectState.setText("连接成功");
                mTvConnectState.setVisibility(View.VISIBLE);
                dismissAllowingStateLoss();
                break;

            default:
                isConnecting = false;
                mTvConnectState.setVisibility(View.GONE);
                dismissAllowingStateLoss();
                break;
        }
    }

    private int dp2px(Context context, int dpValue) {
        if (null == context) {
            return dpValue;
        }

        float scale = context.getResources().getDisplayMetrics().density;
        return (int) (dpValue * scale + 0.5f);
    }

}

File: app/src/main/java/com/darklycoder/wifitool/InputWiFiPasswordDialog.kt
package com.darklycoder.wifitool

import android.content.Context
import android.net.NetworkInfo
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.view.*
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import androidx.fragment.app.DialogFragment
import com.darklycoder.wifitool.lib.WiFiModule
import com.darklycoder.wifitool.lib.info.WiFiScanInfo
import com.darklycoder.wifitool.lib.interfaces.ConnectWiFiActionListener

/**
 * wifi密码输入框
 */
class InputWiFiPasswordDialog : DialogFragment() {

    private var mTvTitle: TextView? = null
    private var mEtPassword: EditText? = null
    private var mTvConnect: TextView? = null
    private var mIvClose: ImageView? = null
    private var mTvConnectState: TextView? = null

    private var mInfo: WiFiScanInfo? = null
    private var isConnecting = false
    private var mConnectType: Int = 0

    private var mConnectActionListener: ConnectWiFiActionListener? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        dialog!!.requestWindowFeature(Window.FEATURE_NO_TITLE)
        return inflater.inflate(R.layout.dialog_input_wifi_password, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val bundle = arguments
        if (null != bundle) {
            mInfo = bundle.getParcelable("info")
            mConnectType = bundle.getInt("connectType", 0)
        }
        if (null == mInfo) {
            dismissAllowingStateLoss()
            return
        }

        mTvTitle = view.findViewById(R.id.tv_title)
        mEtPassword = view.findViewById(R.id.et_password)
        mTvConnect = view.findViewById(R.id.tv_connect)
        mIvClose = view.findViewById(R.id.iv_close)
        mTvConnectState = view.findViewById(R.id.tv_connect_state)

        if (mConnectType == 1) {
            mTvTitle?.text = "修改" + mInfo?.scanResult?.SSID

        } else {
            mTvTitle?.text = mInfo?.scanResult?.SSID
        }

        mEtPassword?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                refreshConnectBtn()
            }

            override fun afterTextChanged(s: Editable) {

            }
        })

        mEtPassword?.isFocusable = true
        mEtPassword?.isFocusableInTouchMode = true
        mEtPassword?.requestFocus()

        mEtPassword?.post {
            try {
                val inputMethodManager = context!!.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                inputMethodManager.toggleSoftInput(0, InputMethodManager.HIDE_NOT_ALWAYS)

            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        mIvClose?.setOnClickListener { mEtPassword?.setText(null) }

        view.findViewById<View>(R.id.tv_cancel).setOnClickListener { dismissAllowingStateLoss() }

        view.findViewById<View>(R.id.tv_connect).setOnClickListener {
            // 连接
            if (isConnecting) {
                return@setOnClickListener
            }

            isConnecting = true

            WiFiModule.connectWiFi(mInfo?.scanResult?.SSID!!, mInfo?.cipherType!!, mEtPassword?.text.toString().trim { it <= ' ' }, mConnectActionListener)

            mTvConnect?.isEnabled = !isConnecting
        }
    }

    override fun onStart() {
        super.onStart()

        val dialog = dialog
        dialog!!.setCanceledOnTouchOutside(true)

        val window = dialog.window
        if (null != window) {
            val lp = window.attributes
            lp.gravity = Gravity.CENTER
            lp.width = dp2px(context, 280)
            lp.height = WindowManager.LayoutParams.WRAP_CONTENT
            window.attributes = lp
        }
    }

    private fun isEmpty(password: String): Boolean {
        return TextUtils.isEmpty(password)
    }

    private fun isValidPassword(password: String): Boolean {
        if (isEmpty(password)) {
            return false
        }

        //6~8位
        val size = password.length
        return size >= 8
    }

    private fun refreshConnectBtn() {
        val password = mEtPassword?.text.toString().trim { it <= ' ' }

        mIvClose?.visibility = if (isEmpty(password)) View.GONE else View.VISIBLE
        mTvConnect?.isEnabled = !isConnecting && isValidPassword(password)
    }

    fun setConnectListener(actionListener: ConnectWiFiActionListener?) {
        this.mConnectActionListener = actionListener
    }

    /**
     * 连接状态回调
     */
    fun connectCallBack(state: NetworkInfo.DetailedState) {
        when (state) {
            NetworkInfo.DetailedState.CONNECTING -> {
                mTvConnectState?.text = "连接中..."
                mTvConnectState?.visibility = View.VISIBLE
            }

            NetworkInfo.DetailedState.DISCONNECTED -> {
                isConnecting = false
                refreshConnectBtn()
                mTvConnectState?.text = "连接失败"
                mTvConnectState?.visibility = View.VISIBLE
            }

            NetworkInfo.DetailedState.CONNECTED -> {
                isConnecting = false
                mTvConnectState?.text = "连接成功"
                mTvConnectState?.visibility = View.VISIBLE
                dismissAllowingStateLoss()
            }

            else -> {
                isConnecting = false
                mTvConnectState?.visibility = View.GONE
                dismissAllowingStateLoss()
            }
        }
    }

    private fun dp2px(context: Context?, dpValue: Int): Int {
        if (null == context) {
            return dpValue
        }

        val scale = context.resources.displayMetrics.density
        return (dpValue * scale + 0.5f).toInt()
    }

}
